File: compiler/nodes/patterns/structure.go
package patterns

import (
	"strconv"

	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type StructurePatternNode struct {
	Name   string
	Fields []StructurePatternField
	Facts  *database.Facts

	matching database.Node

	fieldTemporaries map[string]fieldTemporary
}

type fieldTemporary struct {
	temporary database.Node
	pattern   database.Node
}

func (node *StructurePatternNode) GetFacts() *database.Facts {
	return node.Facts
}

type StructurePatternField struct {
	Name    string
	Pattern database.Node
	Span    database.Span
}

func ParseStructurePattern(parser *syntax.Parser) (*StructurePatternNode, *syntax.Error) {
	span := parser.Spanned()

	name, err := syntax.ParseTypeName(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("LeftBrace")
	if err != nil {
		return nil, err
	}

	fields, err := syntax.ParseLines(parser, 1, true, ParseStructurePatternField)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightBrace")
	if err != nil {
		return nil, err
	}

	return &StructurePatternNode{
		Name:   name,
		Fields: fields,
		Facts:  database.NewFacts(span()),
	}, nil
}

func ParseStructurePatternField(parser *syntax.Parser) (StructurePatternField, *syntax.Error) {
	span := parser.Spanned()

	name, err := syntax.ParseVariableName(parser)
	if err != nil {
		return StructurePatternField{}, err
	}

	_, err = parser.Token("AssignOperator")
	if err != nil {
		return StructurePatternField{}, err
	}

	parser.ConsumeLineBreaks()

	value, err := ParsePattern(parser)
	if err != nil {
		return StructurePatternField{}, err
	}

	return StructurePatternField{
		Name:    name,
		Pattern: value,
		Span:    span(),
	}, nil
}

func (node *StructurePatternNode) Visit(visitor *visit.Visitor) {
	visitPattern(visitor, node)
	node.matching = visitor.CurrentMatch.Node

	definition, ok := visit.Resolve[*visit.StructureConstructorDefinition](visitor, node.Name, node)
	if !ok {
		return
	}

	node.fieldTemporaries = map[string]fieldTemporary{}
	for _, field := range node.Fields {
		node.fieldTemporaries[field.Name] = fieldTemporary{
			temporary: visitor.VisitMatching(field.Pattern),
			pattern:   field.Pattern,
		}
	}

	replacements := map[database.Node]database.Node{definition.Node: node}
	for fieldName, fieldType := range definition.Fields {
		value, ok := node.fieldTemporaries[fieldName]
		if ok {
			replacements[fieldType] = value.temporary
		}
	}

	visitor.Constraint(typecheck.InstantiateConstraint(typecheck.Instantiation{
		Source:        node,
		Definition:    definition.Node,
		Substitutions: &map[database.Node]typecheck.Type{},
		Replacements:  replacements,
	}, visit.GetDefinitionConstraints))
}

func (node *StructurePatternNode) Codegen(c *codegen.Codegen) error {
	if node.fieldTemporaries == nil {
		return c.Error(node)
	}

	span := database.GetSpanFact(node)

	for name, field := range node.fieldTemporaries {
		c.WriteString(span, " && ((")
		c.WriteNode(span, field.temporary)
		c.WriteString(span, " = ")
		c.WriteNode(span, node.matching)
		c.WriteString(span, "[")
		c.WriteString(span, strconv.Quote(name))
		c.WriteString(span, "]) || true)")

		if err := c.Write(field.pattern); err != nil {
			return err
		}
	}

	return nil
}

func (node *StructurePatternNode) EachTemporary(f func(database.Node)) {
	for _, field := range node.fieldTemporaries {
		f(field.temporary)
		EachTemporary(field.pattern, f)
	}
}

File: compiler/src/nodes/patterns/structure.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Fact, Node, NodeRef, Render},
    nodes::{Matching, Temporaries, parse_pattern, visit_pattern},
    syntax::{ParseError, Parser, TokenKind, parse_type_name, parse_variable_name},
    typecheck::{Instantation, InstantiateConstraint, Replacements, Substitutions},
    visit::{Definition, MatchPathSegment, Visit, Visitor},
};

#[derive(Debug)]
pub struct StructurePatternNode {
    pub name: String,
    pub fields: Vec<StructurePatternField>,
}

#[derive(Debug, Clone)]
pub struct StructurePatternField {
    pub name: String,
    pub pattern: NodeRef,
}

#[derive(Debug, Clone)]
struct ResolvedStructurePattern {
    fields: Vec<(Option<usize>, String, NodeRef)>,
}

impl Fact for ResolvedStructurePattern {}

impl Render for ResolvedStructurePattern {}

impl Node for StructurePatternNode {}

pub fn parse_structure_pattern(
    parser: &mut Parser<'_>,
) -> Result<StructurePatternNode, ParseError> {
    let name = parse_type_name(parser)?;
    parser.token(TokenKind::LeftBrace)?;

    let fields = parser
        .parse_lines(1, true, parse_structure_pattern_field)?
        .into_iter()
        .collect();

    parser.token(TokenKind::RightBrace)?;

    Ok(StructurePatternNode { name, fields })
}

pub fn parse_structure_pattern_field(
    parser: &mut Parser<'_>,
) -> Result<StructurePatternField, ParseError> {
    let name = parse_variable_name(parser)?;
    parser.token(TokenKind::AssignOperator)?;
    parser.consume_line_breaks();
    let pattern = parse_pattern(parser)?;

    Ok(StructurePatternField { name, pattern })
}

impl Visit for StructurePatternNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_pattern(node, visitor, None);

        let Some(definition) = visitor.resolve(&self.name, node, |definition| match definition {
            Definition::StructureConstructor(definition) => Some(definition.clone()),
            _ => None,
        }) else {
            return;
        };

        let mut field_temporaries = Vec::new();
        let mut field_values = Vec::new();
        for field in &self.fields {
            let temporary = visitor.visit_matching(
                &field.pattern,
                definition
                    .fields
                    .iter()
                    .find_map(|(name, ty)| (*name == field.name).then_some(ty))
                    .cloned()
                    .map(MatchPathSegment::Field),
            );

            visitor.edge(&field.pattern, node, "field");

            field_temporaries.push((field.name.as_str(), temporary));
            field_values.push(field.clone());
        }

        let definition_node = definition.node.clone();

        let field_values = field_values
            .iter()
            .map(|field| {
                let index = definition
                    .fields
                    .iter()
                    .position(|(name, _)| *name == field.name);

                (index, field.name.clone(), field.pattern.clone())
            })
            .collect::<Vec<_>>();

        let replacements = Replacements::from_iter([(definition_node.clone(), node.clone())]);
        for (field_name, field_type) in &definition.fields {
            if let Some((_, temporary)) = field_temporaries
                .iter()
                .find(|&(name, _)| *name == *field_name)
            {
                replacements.insert(field_type.clone(), temporary.clone());
            }
        }

        visitor.constraint(InstantiateConstraint::new(Instantation {
            source_node: node.clone(),
            definition: definition_node,
            substitutions: Substitutions::new(),
            replacements,
            from_expression: true,
        }));

        visitor.insert(
            node,
            ResolvedStructurePattern {
                fields: field_values,
            },
        );

        visitor.insert(
            node,
            Temporaries {
                has: field_temporaries
                    .into_iter()
                    .map(|(_, temporary)| temporary)
                    .collect(),
                inherit: self
                    .fields
                    .iter()
                    .map(|field| field.pattern.clone())
                    .collect(),
            },
        )
    }
}

impl Codegen for StructurePatternNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        let Temporaries {
            has: field_temporaries,
            ..
        } = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        let Matching(matching) = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        let ResolvedStructurePattern { fields } = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        for ((index, name, pattern), temporary) in fields.iter().zip(field_temporaries) {
            let index = index.ok_or_else(|| anyhow::format_err!("unresolved"))?;

            ctx.condition(ir::Condition::Initialize {
                variable: temporary,
                node: None,
                value: ir::Value::Field {
                    input: matching.clone(),
                    field_name: name.clone(),
                    field_index: index,
                },
                mutable: false,
            });

            ctx.codegen(pattern)?;
        }

        Ok(())
    }
}
